File: #30_Substring_with_Concatenation_of_All_Words/Solution.py
class Solution:
	# @param {string} s
	# @param {string[]} words
	# @return {integer[]}
	def findSubstring(self, s, words):
		if not s or not words:
			return []
		re = []
		map = {}
		for word in words:
			map[word] = map.get(word, 0) + 1
		wlen = len(words[0])
		window = {}
		for x in range(wlen):
			i = x
			j = x
			while j + wlen <= len(s):
				sub = s[j:j + wlen]
				j += wlen
				if sub in map:
					window[sub] = window.get(sub, 0) + 1
				else:
					i = j
					window.clear()
				if j - i == wlen * len(words):
					if window == map:
						re.append(i)
					delStr = s[i: i + wlen]
					window[delStr] = window[delStr] - 1
					if window[delStr] == 0:
						del window[delStr]
					i += wlen
			window.clear()
		return re


sol = Solution()
print sol.findSubstring("barfoothefoobarman", ["foo", "bar"])

File: #30_Substring_with_Concatenation_of_All_Words/Solution.java
import java.util.HashSet;
import java.util.ArrayList;
import java.util.List;

public class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
		ArrayList<Integer> rslt = new ArrayList<Integer>();
		if (words.length == 0) return rslt;
		int len = words[0].length();
		for (String word : words) {
			List<Integer> indices = findOccurance(s, word);
			for (int i : indices) {
				// System.out.println(word + " " + i);
				HashSet<String> set = new HashSet<String>();
				for (String w : words) {
					if (!w.equals(word)) set.add(w);
				}
				search(s.substring(i + len), set, rslt, i, len);
			}
		}
		return rslt;
	}

	private void search(String s, HashSet<String> set, ArrayList<Integer> result, int position, int len) {
		if (set.isEmpty()) return;
		for (String word : set) {
			int index = s.indexOf(word);
			if (index == 0) {
				HashSet<String> subset = new HashSet<String>();
				for (String w : set) {
					if (!w.equals(word)) subset.add(w);
				}
				if (subset.isEmpty()) {
					result.add(position);
					return;
				}
				search(s.substring(len), subset, result, position, len);
			}
		}
	}

	private List<Integer> findOccurance(String s, String word) {
		List<Integer> result = new ArrayList<Integer>();
		if (word.length() > s.length()) return result;
		for (int i = 0; i < s.length() - word.length() + 1; i++) {
			int p = i;
			int q = 0;
			while (s.charAt(p) == word.charAt(q)) {
				p++; q++;
				if (q == word.length()) {
					result.add(i);
					break;
				}
			}
		}
		return result;
	}

	public static void main(String[] args) {
		// String s = "barfoothefoobarman";
		// String[] words = {"foo", "bar"};
		// Solution sol = new Solution();
		// List<Integer> list;
		// list = sol.findSubstring(s, words);
		// for (int i : list) {
		// 	System.out.println(i);
		// }
		for (int i = 0; i < 5; i++) {
			System.out.println(i);
		}
		for (int i = 0; i < 5; ++i) {
			System.out.println(i);
		}

	}
}